core/services/log/pool.go

```diff
@@ -0,0 +1,41 @@
+package log
+
+import (
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+)
+
+type (
+	logPool struct {
+		allLogs []types.Log
+	}
+)
+
+func newLogPool() *logPool {
+	return &logPool{
+		allLogs: make([]types.Log, 0),
+	}
+}
+
+func (pool *logPool) addLog(log types.Log) {
+	pool.allLogs = append(pool.allLogs, log)
+}
+
+func (pool *logPool) getLogsToSend(head *models.Head, highestNumConfirmations uint64) []types.Log {
+	latestBlockNum := uint64(head.Number)
+	logsToReturn := pool.allLogs
+	logsToKeep := make([]types.Log, 0)
+
+	// deleting old logs that will never be sent for any listener anymore
+	if latestBlockNum > highestNumConfirmations && len(pool.allLogs) > 0 {
+		for _, log := range pool.allLogs {
+			if log.BlockNumber >= latestBlockNum-highestNumConfirmations {
+				logsToKeep = append(logsToKeep, log)
+			}
+		}
+		logger.Tracef("latestBlockNum (%v) > highestNumConfirmations (%v) so deleting older logs. Remaining: %v", latestBlockNum, highestNumConfirmations, len(logsToKeep))
```

Comment:
Personally I prefer structured logging `logger.Tracew` but you can combine both with `fmt.Sprintf`

Reply:
yeah, can change to Tracew - i wasn't sure if we are going to keep those traces though

Reply:
I tend to favour more logging over less, especially since this is at Trace level.